src/sdglib/factories/Kmers.h

```diff
@@ -0,0 +1,181 @@
+//
+// Created by Ben J. Ward (EI) on 27/05/2020.
+//
+
+#ifndef SDG_KMERS_H
+#define SDG_KMERS_H
+
+#pragma once
+
+#include <array>
+#include <limits>
+#include <assert.h>
+#include <iostream>
+
+// I want this function to be able to run at compile time to create the constant static
+// arrays that the Kmers class uses to convert chars to binary.
+template<typename U>
+constexpr std::array<U, 256> make_nuc2bin_tbl(){
+    std::array<U, 256> tbl = {4};
+    tbl['a'] = tbl['A'] = 0;
+    tbl['c'] = tbl['C'] = 1;
+    tbl['g'] = tbl['G'] = 2;
+    tbl['t'] = tbl['T'] = 3;
+    return tbl;
+}
+
+// A type representing a "lazy" - if you want to call it that, container of all the k-mers in a sequence.
+// Has an input iterator in its scope for iterating from the first k-mer in a sequence to the last.
+template <typename U> class Kmers {
+protected:
+    // C++11 (explicit aliases for input iterator).
+    using iterator_category = std::input_iterator_tag;
+    using value_type = U;
+    using reference = value_type const&;
+    using pointer = value_type const*;
+    using difference_type = ptrdiff_t;
+
+public:
+    // Unwrap std::string and pass to the cstring constructor.
+    Kmers(const std::string& seq, const int K) : Kmers(seq.data(), K) {} // Requires C++11.
+
+    Kmers(const char * seq, const int K) : sequence(seq), K(K) {
+        // Here I check someone hasn't requested silly K size, before
+        // making the local translation table for this container.
+        // static asserts below?
+        assert(K > 0);
+        assert(!std::numeric_limits<value_type>::is_signed);
+        // Forgive heavy bracket use - I need to get used to if C++ has the operator precedence I'm used to or not.
+        assert(K <= ((std::numeric_limits<value_type>::digits / 2))); // -1 here to keep space for the sign bits.
+        translation_table = make_nuc2bin_tbl<value_type>();
+    }
+
+    class iterator: public std::iterator<iterator_category, value_type, difference_type, pointer, reference> {
+    public:
+        explicit iterator(Kmers& p) : parent(p) {
+            position = parent.sequence;
+            last_unknown = 0;
+            fwmer = 0;
+            rvmer = 0;
+            mask = (value_type(1) << (2 * parent.K)) - 1;
+            incorporate_char(position);
+        }
+        bool has_reached_end() { return *position == '\0'; }
+
+        reference operator*() const {
+            // I want to specialize this on template parameter F but am unsure on how to do it properly.
+            return std::min(fwmer & mask, rvmer & mask); // SEGFAULTS if I make it non-canonical - WTF.
+        }
+        pointer operator->() const {
+            // I want to specialize this on template parameter F but am unsure on how to do it properly.
+            return fwmer & mask;
+        }
```

Comment:
is this the right type and return for this ?

Reply:
I'm not sure, I don't really get this operator, because I don't get what (it->) would mean without a field specified (it->mer or soemthing), but tutorials on making iterators define it like this.

Reply:
Deleted this and let C++ just do whatever is default.